src/gridcoin/voting/registry.cpp

```diff
@@ -311,6 +317,174 @@ int64_t PollReference::Expiration() const
+    // pass, which is the most efficient.
+    double active_vote_weight_tally = 0;
+    double pool_magnitude = 0.0;
+    double network_magnitude = 0.0;
+    unsigned int blocks = 0;
+
+    // Lambda for active_vote_weight tally.
+    const auto tally_active_vote_weight = [&](
+            const double net_weight,
+            const double money_supply,
+            const double pool_magnitude,
+            const double network_magnitude)
+    {
+        active_vote_weight_tally += net_weight + money_supply * (network_magnitude - pool_magnitude)
+                                                              / (5.67 * network_magnitude);
+
+        ++blocks;
+    };
+
+    // Rewind from pindex_start to find last superblock before start of the poll to pick up first pool magnitudes
+    for (CBlockIndex* pindex = pindex_start; pindex; pindex = pindex->pprev)
+    {
+        if (pindex->IsSuperblock()) {
+            const GRC::ClaimOption claim = GetClaimByIndex(pindex);
+
+            // Add up the magnitudes of the pools in the last superblock before the start of the poll.
+            if (claim && claim->ContainsSuperblock()) {
+                const GRC::Superblock& superblock = *claim->m_superblock;
+
+                for (const auto& pool : g_mining_pools.GetMiningPools()) {
+                    pool_magnitude += superblock.m_cpids.MagnitudeOf(pool.m_cpid).Floating();
+                }
+
+                network_magnitude = (double) superblock.m_cpids.TotalMagnitude();
+            }
+
+            // Stop after processing the first superblock found rewinding from the poll start.
+            break;
+        }
+    }
+
+    // Scan the index for the poll duration and compute AVW.
+    for (CBlockIndex* pindex = pindex_start; pindex; pindex = pindex->pnext) {
+        // Refresh pool magnitude and network magnitude if the index points to a superblock. The pool_magnitude and
+        // network magnitude remain constant for all subsequent blocks until replaced by the values from a fresh superblock
+        // in the scan.
+        if (pindex->IsSuperblock()) {
+            pool_magnitude = 0.0;
+            network_magnitude = 0.0;
+
+            const GRC::ClaimOption claim = GetClaimByIndex(pindex);
+
+            // Add up the magnitudes of the pools in the last superblock before the start of the poll.
+            if (claim && claim->ContainsSuperblock()) {
+                const GRC::Superblock& superblock = *claim->m_superblock;
+
+                for (const auto& pool : g_mining_pools.GetMiningPools()) {
+                    pool_magnitude += superblock.m_cpids.MagnitudeOf(pool.m_cpid).Floating();
+                }
+
+                network_magnitude = (double) superblock.m_cpids.TotalMagnitude();
+            }
+        }
+
+        // Please refer to https://gridcoin.us/assets/docs/grc-bluepaper-section-1.pdf equation 16.
+        double net_weight = 9544517.4 * GetDifficulty(pindex);
```

Comment:
https://github.com/gridcoin-community/Gridcoin-Research/blob/d66f81ab72099a06e215fba2f68937bd1555cba7/src/gridcoin/staking/difficulty.cpp#L265
It seems that the constant `GRC::GetEstimatedNetworkWeight` is calculated with the slightly more precise `9544517.40667`. They both should use the same value in my opinion.

Nits related to the poll proposal:
On a related note, the constant calculated in the bluepaper is different than the one accepted in the poll proposal poll. Assuming that the one in bluepaper is more correct, it is not a big problem in my opinion.

"Pool Magnitude Sum" makes more sense than "Average Pool Magnitude"  (as implemented here).

Unconditionally subtracting is not the best idea in my opinion when pools could give the users the ability to vote(I remember grcpool doing this?).

Reply:
> "Pool Magnitude Sum" makes more sense than "Average Pool Magnitude" (as implemented here).

Average means average over the duration of the poll and not (pool1mag + pool2mag)/2 or similar 

> Unconditionally subtracting is not the best idea in my opinion when pools could give the users the ability to vote(I remember grcpool doing this?).

This is what the network approved definition of AVW does. It simply subtracts the pool magnitudes. Diverging from that could be problematic 

Grcpool never actually voted on the blockchain - the vote tab there was just to get an informal sense of what pool users thought

Reply:
1. The constant value in the bluepaper is the one that should be used. It is the correct value. I will use the more precise version in the GetEstimatedNetworkWeight (which I wrote a long time ago anyway).
2. The tally approach to the averaging is the most correct way (we will discuss integer vs floating point math below). Look carefully at the notes I made above. In particular...
![AVW_calcs-1](https://user-images.githubusercontent.com/7529186/124635812-dd38b680-de55-11eb-928a-5248f6ab20fd.jpg)
![AVW_calcs-2](https://user-images.githubusercontent.com/7529186/124635813-ddd14d00-de55-11eb-94dc-63c025d64e8b.jpg)

Reply:
On the underflow, from the logging...
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 13066910.689778, money_supply = 394969597.921623, pool_magnitude = 0.000000, network_magnitude = 0.000000, blocks = 1, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 15097025.193305, money_supply = 395565213.108492, pool_magnitude = 25600.000000, network_magnitude = 113748.000000, blocks = 16436, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 24834938.813607, money_supply = 395634501.853444, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18232, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 28131109.631737, money_supply = 395634560.059292, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18233, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 31864763.122437, money_supply = 395634568.627231, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18234, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 30508820.793263, money_supply = 395634636.571351, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18235, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 30143446.553697, money_supply = 395634720.571451, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18236, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z INFO tally_active_vote_weight: net_weight = 27178520.236160, money_supply = 395634785.425937, pool_magnitude = 25520.000000, network_magnitude = 112898.000000, blocks = 18237, active_vote_weight_tally = -nan, active_vote_weight = -nan.
2021-07-06T16:51:07Z timer GetActiveVoteWeight: finished execution for poll id df3af5f55268bd71b7498da4690d36b9ba7eb1f403162f72fce5d3cd950e04d9 title technical_poll:_should_we_implement_fixed_pos_block_rewards?: elapsed time: 177 ms, time since last check: 177 ms.

Looks like the first superblock is not being picked up properly. Let me look at that.

Reply:
And by the way, this is why you cannot assume that network magnitude is actually 115000, because charity engine and the pools were capped to 16383 and remember that renormalization does not occur when CPID magnitude caps are reached. The cap was doubled to 32767, so this rarely happens these days.

Reply:
Dang it. On the underflow, looks like the IsSuperblock() is broken for superblocks before 1034768. That is causing the underflow.
See...
[superblocks.log](https://github.com/gridcoin-community/Gridcoin-Research/files/6771744/superblocks.log)

Reply:
Workaround implemented that appears to work.